yt_dlp/utils/_utils.py

```diff
@@ -1361,6 +1361,23 @@ def brotli(data):
             return data
         return brotli.decompress(data)
 
+    @staticmethod
+    def gz(data):
+        gz = gzip.GzipFile(fileobj=io.BytesIO(data), mode='rb')
+        try:
+            return gz.read()
+        except OSError as original_ioerror:
+            # There may be junk add the end of the file
+            # See http://stackoverflow.com/q/4928560/35070 for details
+            for i in range(1, 1024):
+                try:
+                    gz = gzip.GzipFile(fileobj=io.BytesIO(data[:-i]), mode='rb')
+                    return gz.read()
+                except OSError:
+                    continue
```

Comment:
Could you introspect the error and see if we have access to the required offset? I don't like the idea of re-reading in a loop if it can be helped.

Reply:
Sadly that doesn't really work:
```py
>>> import gzip
>>> data = gzip.compress(b"01234567890123456789asdfasdfasdf")
>>> data += b"11111"
>>> gzip.decompress(d)
gzip.BadGzipFile: Not a gzipped file (b'11')
```

Reply:
Hmm.

This raises BadGzipFile (subclass of OSError  in 3.8) with the message `Not a gzipped file (b'tr')`. No other useful attributes are set.

`tr` in this case is the start of the trailing garbage in the raw response content: `b'\x1f\x8b\x08\x00u\xd1qd\x02\xff\xb3\xc9(\xc9\xcd\xb1\xb3)\xcbLI\xcdW(.J\xb6U\xd2\x07\xb2\xf5r\x0bL\x94\x14\xf4\xedl\xf4\xc1\xd2\x00R)\x10\x96%\x00\x00\x00trailing garbage'`

Reply:
Alternative idea: `zlib.decompress(data, wbits=31)`, it seems to ignore trailing data?

Reply:
> Alternative idea: `zlib.decompress(data, wbits=31)`, it seems to ignore trailing data?

could be just a coincidence

Reply:
The original `gzip` version tries to decode the "trailing data" as another payload, as specified by the error message (Not a gzipped file). `zlib` instead only decodes one payload, which is why it shouldn't fail for trailing data. I've tried with various different gzipped payloads and various sizes of trailing data.

Reply:
Ah just saw your edit. Hmm.

Having a look at what `urllib3` is doing, they are using `zlib` too. Bit with some more processing for trailing data:

https://github.com/urllib3/urllib3/blob/main/src/urllib3/response.py#L119-L139:

Reply:
might be better to leave this to another PR/commit if its something you may want to work on.

as it will likely require its own set of reviews and testing

Reply:
Tests should maybe be expanded to have more trailing data or sth, but otherwise, wouldn't it be just
```diff
    @staticmethod
    def gz(data):
-        gz = gzip.GzipFile(fileobj=io.BytesIO(data), mode='rb')
-        try:
-            return gz.read()
-        except OSError as original_oserror:
-            # There may be junk add the end of the file
-            # See http://stackoverflow.com/q/4928560/35070 for details
-            for i in range(1, 1024):
-                try:
-                    gz = gzip.GzipFile(fileobj=io.BytesIO(data[:-i]), mode='rb')
-                    return gz.read()
-                except OSError:
-                    continue
-            else:
-                raise original_oserror
+        return zlib.decompress(data, 31)
```
We already have the whole data as `bytes`, so not fitting into memory should not be a problem. If you want I can open a new PR though

Reply:
yeah better as its own PR if you want to. I'll merge this one